### ISSUE TITLE
Replica gets stuck in intermediate status if client process is terminated

### ISSUE THREAD

**Issue by mstfdkmn** [irods/irods_client_http_api#188]
### iRODS Version, OS and Version
4.3.1, almalinux8

### What did you try to do?
While writing I cut the signal (ctr+c).

### Expected behavior
The object could be in stale `X` for some moment or immediately or a bit later in good `&`

### Observed behavior (including steps to reproduce, if applicable)
The object gets locked.

```
[irods@dev-u0137480 go-http-client]$ time curl -H "Authorization: Bearer $token" 'http://localhost:9000/irods-http-api/0.1.0/data-objects' -F  'op=write' -F 'lpath=/u0137480/home/rods/myfile.txt' -F 'count=125000' -F 'bytes=abcdfg'
^C

real    0m3.778s
user    0m0.003s
sys     0m0.006s
[irods@dev-u0137480 go-http-client]$ ils -L /u0137480/home/rods/myfile.txt
  rods              0 default            0 2023-12-22.17:06 ? myfile.txt
        generic    /vault/home/rods/myfile.txt

```

**Comment by trel** [comment]
I believe once the server connection times out, it will shift to `X`, or stale.   But the server probably was not told the client connection went away.  We will work to reproduce.

**Comment by korydraughn** [comment]
That's correct. The operation does not know the client disappeared, which isn't incorrect. Perhaps the client doesn't care about the response. Of course, that makes resolving this issue slightly more difficult.

What if the HTTP API provided a way for admins to view active transfers? Hmm ...

Again, the behavior is unknown due to the mismatch values of the `count` and `bytes` params.

**Comment by trel** [comment]
Right.  So, once we resolve the mismatch scenario... this particular issue will become more clear.

**Comment by korydraughn** [comment]
The parameter mismatch issue has been resolved by #194.

Even with the parameter issue resolved, I think we may want to provide a way to cancel a transfer. Which implies we have to provide a way to view active transfers.

Thoughts?

**Comment by trel** [comment]
Who would be able to cancel a transfer?   Presumably just admins.

And for what scenario can they not just ... wait for timeout?

Not sure the added complexity / states are worth the effort.

Do any other HTTP services provide these kinds of insight/listing/cancel button?

**Comment by korydraughn** [comment]
> Who would be able to cancel a transfer? Presumably just admins.

Admins would be able to view all transfers. Non-admins would be able to view just their transfers.
 
> And for what scenario can they not just ... wait for timeout?

With #194 handled, all transfers should end eventually.

This idea would help in situations where the amount of data being transferred is very large.

> Not sure the added complexity / states are worth the effort.
> 
> Do any other HTTP services provide these kinds of insight/listing/cancel button?

I'm not sure what other HTTP APIs provide, but this idea doesn't sound out of bounds for iRODS. I agree the added complexity may not be worth the effort though.

I'm convinced we should wait until more users appear before adding something like this.

**Comment by trel** [comment]
Perhaps update the title to be... 'provide status and cancellation API for running transfers'

**Comment by korydraughn** [comment]
> And what if the reason of the fail would be a crash of the server? HTTP API would lose the handles from the memory and object will be locked without possibility to unlock. Or do you have separated DB for HTTP API?

When you "_the server_", are you referring to the HTTP API server or the iRODS server?

Assuming you're referring to the HTTP API server - if it crashes, the iRODS server auto-closes all replicas tied to that agent.

If you're wondering whether `rc_switch_user` would lead to issues _(like me)_, there's no need to be concerned with that since the HTTP API always sets the option to close open replicas **before** changing the identity associated with the connection.

So, on a crash, open replicas will be marked stale if the iRODS server is healthy.

---

GenQuery gives us the ability to see intermediate replicas. That's good. What we're missing is the ability to track who opened what _(or maybe we already capture that?)_. That is information which should be stored in the catalog. If we had that info, we could develop a nice solution that works for all clients.

I think that piece of info could be managed by the logical locking system and finalize API.

**Comment by trel** [comment]
After some discussion, we are proposing the following two-part solution...

1) in the iRODS Server
 - probably also solves https://github.com/irods/irods/issues/8765
 - when locks are created, also save the username (or perhaps the user_id) in `r_data_main.data_status` for each affected replica
   - perhaps as `user_id=xxxx`, in case we ever decide to add something else to this later.  This would introduce the requirement for a parser and/or make matching via genquery have to be a little smarter (using `DATA_STATUS LIKE 'user_id=%;', requiring the semicolon to delimit to avoid `2000` matching `20000`)
   - perhaps the status field ALSO holds 'original status' or other information that was kept in the replica state table in memory (and which is presumably already lost/gone)
 - create a new API, probably `unlock_replica`, or perhaps just `stale_replica`.
   - `unlock_replica(logical_path, replica_number)`
   - could potentially alternatively take data_id and/or resc_hier, like `iadmin modrepl`
   - errors if the calling user does not own the data object
   - errors if the calling user does not match the user stored in `r_data_main.data_status`
   - potential additional checks... servername, pid, magic token... but probably not


2) in the HTTP API (or ANY client)...
 - when either `INTERMEDIATE_REPLICA_ACCESS (2)` or `LOCKED_DATA_OBJECT_ACCESS (4)` is returned by the server suggesting the target is locked...
   - genquery the owner of the lock (`DATA_STATUS`), if it matches...
     - then 'fix' the status via the new `unlock_replica()` API, and if success...
       - continue overwriting the target object


If/when clients introduce this 'best practice' client-side algorithm to avoid waiting for the server to timeout and unlock the replica itself, then it re-opens the possibility of uncoordinated client writers overwriting current work (but only if they are by the same user and pointing to the same data object).  This would be a bad design in the client side application and would be no different than writing into a regular filesystem without uniqueness checks.
